src/Events.Functions/Services/CertificateResolverService.cs

```diff
@@ -0,0 +1,75 @@
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+using Altinn.Platform.Events.Functions.Configuration;
+using Altinn.Platform.Events.Functions.Services.Interfaces;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Altinn.Platform.Events.Functions.Services
+{
+    /// <summary>
+    /// Class to resolve certificate to generate an access token
+    /// </summary>
+    /// <remarks>This class is excluded from code coverage because it has no logic to be tested.</remarks>
+    [ExcludeFromCodeCoverage]
+    public class CertificateResolverService : ICertificateResolverService
+    {
+        private readonly ILogger<ICertificateResolverService> _logger;
+        private readonly CertificateResolverSettings _certificateResolverSettings;
+        private readonly IKeyVaultService _keyVaultService;
+        private readonly KeyVaultSettings _keyVaultSettings;
+        private DateTime _expiryTime;
+        private X509Certificate2 _cachedX509Certificate = null;
+        private readonly object _lockObject = new object();
+
+        /// <summary>
+        /// Default constructor
+        /// </summary>
+        /// <param name="logger">The logger</param>
+        /// <param name="certificateResolverSettings">Settings for certificate resolver</param>
+        /// <param name="keyVaultService">Key vault service</param>
+        /// <param name="keyVaultSettings">Key vault settings</param>
+        public CertificateResolverService(
+            ILogger<ICertificateResolverService> logger,
+            IOptions<CertificateResolverSettings> certificateResolverSettings,
+            IKeyVaultService keyVaultService,
+            IOptions<KeyVaultSettings> keyVaultSettings)
+        {
+            _logger = logger;
+            _certificateResolverSettings = certificateResolverSettings.Value;
+            _keyVaultService = keyVaultService;
+            _keyVaultSettings = keyVaultSettings.Value;
+            _expiryTime = DateTime.MinValue;
+        }
+
+        /// <summary>
+        /// Find the configured 
+        /// </summary>
+        /// <returns></returns>
+        public async Task<X509Certificate2> GetCertificateAsync()
+        {
+            if (_expiryTime > DateTime.UtcNow && _cachedX509Certificate != null)
+            {
+                return _cachedX509Certificate;
+            }
+
+            string certBase64 = await _keyVaultService.GetCertificateAsync(
```

Comment:
What do you think about moving this statement behind the lock?

Reply:
Ah, okay, yeah. I've been using SemaphoreSlim in other cases. We don't need to change this now. Can look into it in another context.